---
title: "Figure1_Analysis"
author: "Priya Veeraraghavan"
date: "2023-03-25"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warnings=FALSE)
```

```{r}
library(rjson)
input <- "C:/Users/prv892/Documents/GitHub/amalgam/metadata.json"

file_path_ref <- fromJSON(file = input)

# import the relevant parameters and references
source(file.path(file_path_ref$project_directory, 
                 file_path_ref$sequencing$setup_script))

source(file.path(file_path_ref$project_directory,
                 file_path_ref$plots$setup_script))
```

# Main figure
```{r, fig.width=4.6, fig.height=3.3, fig.fullwidth=FALSE}
plot_main_1d <- function(plot_savefile) {
  load(file.path(file_path_ref$project_directory,
                 file_path_ref$sequencing$results$all_dpc23_comparison))
  
  pca_res <- get_pcs(assay(vst_subset))

  plt <- ggplot(pca_res$pcs, 
              aes(PC1, PC2, 
                  color=subtype, shape=location)) + 
         geom_point(size=0.5) +
         xlab(paste0("PC1: ", pca_res$variances[1], "% Variance Explained", collapse="")) +
         ylab(paste0("PC2: ", pca_res$variances[2], "% Variance Explained", collapse="")) +
         theme_classic() + 
    scale_color_manual(values=c(cpn_color, cthpn_color)) +
    theme(legend.position="None",
          text=element_text(size=6))
  
  plot(plt)
  
  
  ggsave(plot_savefile, plt, device="pdf", useDingbats=FALSE, width=46, height=33, units="mm")
  
}

plot_main_1d(file.path(file_path_ref$project_directory,
                       file_path_ref$plots$plot_directory, "main_1d.pdf"))
```

The genes chosen for labeling have previously been described as being localized
and/or with localization elements already determined

```{r, fig.width=5.3, fig.height=5.3, fig.fullwidth=FALSE}
plot_main_1e <- function(plot_savefile) {
  load(file.path(file_path_ref$project_directory,
                              file_path_ref$sequencing$results$gc_vs_soma_dpc23_comparison))
  
  res_dge_valid <- subset(res_dge, is_valid_gc & baseMean > 100) 
  
  res_dge_gc_enriched <- subset(res_dge_valid, 
                                log2FoldChange > 0 & padj < 0.05)
  
  res_dge_soma_enriched <- subset(res_dge_valid, 
                                  log2FoldChange < 0 & padj < 0.05)
  
  res_dge_tolabel <- subset(res_dge_valid, 
                            external_gene_name %in% c("Actb", "Nrn1", "Ybx1", "Creb1") |
                              log2FoldChange > 3)
  

  plt <- rasterize(ggplot(res_dge_valid, 
         aes(baseMean, log2FoldChange)) +
    geom_point(color=background_color,
               size=0.5),
    dpi=300) +
    rasterize(geom_point(data=res_dge_gc_enriched,
                         mapping=aes(baseMean, log2FoldChange),
                         color=accent_color_dark,
                         size=0.5),
              dpi=300) +
      rasterize(geom_point(data=res_dge_soma_enriched,
                         mapping=aes(baseMean, log2FoldChange),
                         color=background_color_dark,
                         size=0.5),
              dpi=300) +
    geom_text_repel(data=res_dge_tolabel,
               mapping=aes(baseMean, log2FoldChange, 
                           label=external_gene_name),
               color="black", size=1) +
    scale_x_log10() +
    theme_classic() +
    ylab("Log2 Fold Change GC/Soma") + 
    xlab("Mean Normalized Counts") +
    theme(text=element_text(size=6))
  
  plot(plt)
  
  ggsave(plot_savefile, plot = plt, 
         width = 53, height = 53, units="mm", 
         useDingbats=FALSE)
  
  
  
  
  
}

plot_main_1e(file.path(file_path_ref$project_directory,
                       file_path_ref$plots$plot_directory, "main_1e.pdf"))
```


```{r, fig.width=6.1, fig.height=6.1, fig.fullwidth=FALSE}
plot_main_1f <- function(plot_savefile) {
  
  load(file.path(file_path_ref$project_directory,
                              file_path_ref$sequencing$results$gc_vs_soma_dpc23_comparison))
  
  go_grouped_simple <- simplify(go_grouped)
  goterm_df <- subset(go_grouped_simple@compareClusterResult, Count >= 10 ) %>%
    rowwise() %>%
    mutate(gene_ratio=gene_ratio_compute(GeneRatio),
           description=str_wrap(Description, width=20)) %>%
    arrange(desc(Cluster), desc(gene_ratio))
  
  goterm_df$description <- factor(goterm_df$description, 
                                  levels=goterm_df$description)
  
  plt <- ggplot(goterm_df, aes(str_wrap(gsub("_", " ", Cluster), width=8), description, size=gene_ratio, color=p.adjust)) +
    geom_point() +
    scale_color_gradient(low=accent_color_go_dark, high=accent_color_go_light)+
    theme_bw() +
    xlab("") + ylab("") +
    theme(text=element_text(size=8))
  
  plot(plt)
  
  ggsave(plot_savefile, plot=plt, device="pdf",
         width=29*3, height=61*3, units="mm", useDingbats=FALSE)
  
}

plot_main_1f(file.path(file_path_ref$project_directory,
                       file_path_ref$plots$plot_directory, "main_1f.pdf"))
```

# Supplement
```{r}
plot_supplement_4b_ambient_ma <- function(plot_saveprefix) {
  
  load(file.path(file_path_ref$project_directory, 
                 file_path_ref$sequencing$results$all_dge_results))
  
  
  gc_vs_gcf <- dge_results$gene_level$gc_vs_gcf
  cpn_colored <- subset(gc_vs_gcf, CPN_GC_DPC23)
  cth_colored <- subset(gc_vs_gcf, CThPN_GC_DPC23)
  
  # plot cpn
  plt1 <- rasterize(ggplot(gc_vs_gcf, 
                           aes(baseMean.CPN_GC_DPC23.vs.Forebrain_Background, 
                               log2FoldChange.CPN_GC_DPC23.vs.Forebrain_Background)) +
    geom_point(color=background_color, size=0.5), dpi=300) +
    rasterize(geom_point(data=cpn_colored, 
                         mapping=aes(baseMean.CPN_GC_DPC23.vs.Forebrain_Background, 
                                     log2FoldChange.CPN_GC_DPC23.vs.Forebrain_Background),
                         color=cpn_color,
                         size=0.5),
              dpi=300) +
    scale_x_log10() +
    ggtitle(sprintf("%s CPN_GC Genes", nrow(cpn_colored))) +
    ylab("Log2 CPN GC/Input") +
    xlab("Average Counts") +
    theme_classic() +
    theme(text=element_text(size=7))
  
  plot(plt1)
  
  plt2 <- rasterize(ggplot(gc_vs_gcf, 
                           aes(log2FoldChange.CThPN_GC_DPC23.vs.Forebrain_Background,
                               log2FoldChange.CThPN_GC_DPC23.vs.Thalamus_Background)) +
                      geom_point(color=background_color, size=0.5), dpi=300) +
    rasterize(geom_point(data=cth_colored,
                         mapping=aes(log2FoldChange.CThPN_GC_DPC23.vs.Forebrain_Background,
                               log2FoldChange.CThPN_GC_DPC23.vs.Thalamus_Background),
                         color=cthpn_color,
                         size=0.5),
              dpi=300) +
        ggtitle(sprintf("%s CThPN_GC Genes", nrow(cth_colored))) +
    ylab("Log2 CTh GC/GCF Thalamus") +
    xlab("Log2 CTh GC/GCF Forebrain")+
    theme_classic() +
    theme(text=element_text(size=7))

  
  plot(plt2)
  
  
  ggsave(sprintf("%s_cpn.pdf", plot_saveprefix),
         plot=plt1, height=70, width=90, units="mm", device="pdf")
  
  ggsave(sprintf("%s_cth.pdf", plot_saveprefix),
         plot=plt2, height=70, width=70, units="mm", device="pdf")
  
  
}

plot_supplement_4b_ambient_ma(file.path(file_path_ref$project_directory,
                                        file_path_ref$plots$plot_directory,
                                        "supplement_4b_ambient_ma"))
```
Option to use Table S4 from single cell mouse brain atlas at P2 and P11 from Rosenberg, et al. 2020, Science (DOI: 10.1126/science.aam8999) in which the top 50 genes from each tSNE cluster are annotated. 
```{r}
plot_contaminants <- function(contam_df, gc_vs_gcf) {
  
  gc_vs_gcf_contam <- dge_results$gene_level$gc_vs_gcf %>%
    inner_join(contam_df, by="external_gene_name") 
  
  median_lfc <- gc_vs_gcf_contam %>%
    group_by(Cell_Type) %>%
    summarize(med_lfc=median(na.omit(log2FoldChange.CPN_GC_DPC23.vs.Forebrain_Background))) %>%
    arrange(desc(med_lfc))
  
  gc_enrich_genes <- subset(dge_results$gene_level$gc_vs_gcf, CPN_GC_DPC23) %>%
    rowwise() %>%
    mutate(Cell_Type="GC_Enriched")
  
  to_plot <- rbind(gc_vs_gcf_contam, gc_enrich_genes) %>%
    rowwise() %>%
    mutate(group_color=ifelse(Cell_Type == "GC_Enriched", "Signal",
                              "Background"))
  to_plot$Cell_Type <- factor(to_plot$Cell_Type, levels=c("GC_Enriched", median_lfc$Cell_Type))
  
  plt <- ggplot(to_plot, aes(Cell_Type, 
                             log2FoldChange.CPN_GC_DPC23.vs.Forebrain_Background,
                             fill=group_color)) +
    geom_boxplot(linewidth=0.2, outlier.size = 0.2) +
    coord_flip() +
    theme_classic() +
    scale_fill_manual(values=c(background_color, cpn_color)) + 
    theme(text=element_text(size=7), legend.position="None")
  
  return(plt)
  
  
}


plot_supplement_4c_ambient_markers <- function(plot_saveprefix) {
  
   load(file.path(file_path_ref$project_directory, 
                 file_path_ref$sequencing$results$all_dge_results))
  
  
  contaminant_small <- read_xlsx(file.path(file_path_ref$project_directory,
                                      file_path_ref$external_data$contaminant_marker_genes)) %>%
    dplyr::select(-Source)
  
  plt1 <- plot_contaminants(contaminant_small)
  ggsave(sprintf("%s_small.pdf", plot_saveprefix),
         plot=plt1,
         height=80, width=80, units="mm", device="pdf")
    
  sc_clust <- read_xlsx("C:/Users/prv892/Documents/GitHub/amalgam/data/external_data/Rosenberg_etal_SingleCell_Clusters.xlsx", skip=1) %>%
    pivot_longer(everything(),
                 names_to="Cell_Type",
                 values_to="external_gene_name")
  
  
  plt2 <- plot_contaminants(sc_clust) 
  
    ggsave(sprintf("%s_single_cell_clusters.pdf", plot_saveprefix),
         plot=plt2,
         height=150, width=120, units="mm", device="pdf")
  
}

plot_supplement_4c_ambient_markers(file.path(file_path_ref$project_directory,
                                             file_path_ref$plots$plot_directory,
                                             "supplement_4c_ambient_markers"))
```

```{r}
plot_supplement_4d_ambient_goterms <- function(plot_savefile) {
  load(file.path(file_path_ref$project_directory, 
                 file_path_ref$sequencing$results$all_dge_results))
  

  gc_enrich <- unique(subset(dge_results$gene_level$gc_vs_gcf, 
                            CPN_GC_DPC23)$ensembl_gene_id)
  input_enrich <- unique(subset(dge_results$gene_level$gc_vs_gcf,
                log2FoldChange.CPN_GC_DPC23.vs.Forebrain_Background < 0 &
                   log2FoldChange.CThPN_GC_DPC23.vs.Forebrain_Background < 0 &
                  log2FoldChange.CThPN_GC_DPC23.vs.Thalamus_Background < 0)$ensembl_gene_id)
  
  bkg <- unique(dge_results$gene_level$gc_vs_gcf$ensembl_gene_id)
  
  go_enrich_bkg <- compareCluster(list(GC_Enriched=gc_enrich,
                      Input_Enriched=input_enrich),
                 ont="ALL", 
                 OrgDb=org.Mm.eg.db,
                 keyType="ENSEMBL",
                 universe=bkg)
  
  plt <- plot_goterm_dotplot(go_enrich_bkg@compareClusterResult) +
    scale_color_gradient(low=accent_color_go_dark, high=accent_color_go_light)
  
  plot(plt)
  
   ggsave(plot_savefile,
         plot=plt,
         height=120, width=100, units="mm", device="pdf",
         useDingbats=FALSE)
  
  
  
}
plot_supplement_4d_ambient_goterms(file.path(file_path_ref$project_directory,
                                             file_path_ref$plots$plot_directory,
                                             "supplement_4d_ambient_goterms.pdf"))
```